src/ParallelAlgorithms/Amr/Actions/AdjustDomain.hpp

```diff
@@ -101,44 +104,70 @@ struct AdjustDomain {
       }
 
     } else {
+      // Determine new neighbors and update the Element
+      {  // avoid shadowing when mutating flags below
+        const auto& amr_flags_of_neighbors =
+            db::get<amr::Tags::NeighborFlags<volume_dim>>(box);
+        db::mutate<::domain::Tags::Element<volume_dim>>(
+            make_not_null(&box),
+            [&element_id, &amr_flags_of_neighbors](
+                const gsl::not_null<Element<volume_dim>*> element) {
+              auto new_neighbors = element->neighbors();
+              for (auto& [direction, neighbors] : new_neighbors) {
+                neighbors.set_ids_to(amr::new_neighbor_ids(
+                    element_id, direction, neighbors, amr_flags_of_neighbors));
+              }
+              *element = Element<volume_dim>(element_id, new_neighbors);
+            });
+      }
       // Check for p-refinement
       if (alg::any_of(my_amr_flags, [](amr::Flag flag) {
             return (flag == amr::Flag::IncreaseResolution or
                     flag == amr::Flag::DecreaseResolution);
           })) {
+        const Mesh<volume_dim> old_mesh =
+            db::get<::domain::Tags::Mesh<volume_dim>>(box);
+
         db::mutate<::domain::Tags::Mesh<volume_dim>>(
             make_not_null(&box),
-            [&my_amr_flags](const gsl::not_null<Mesh<volume_dim>*> mesh) {
-              *mesh = amr::projectors::mesh(*mesh, my_amr_flags);
+            [&old_mesh,
+             &my_amr_flags](const gsl::not_null<Mesh<volume_dim>*> mesh) {
+              *mesh = amr::projectors::mesh(old_mesh, my_amr_flags);
+            });
+
+        // The current implementation assumes that:
+        // -- any mutated items does not need access to the old Element.
+        // -- no item besides the Element needs to be mutated if the Neighbors
+        //    of the Element have changed.
```

Comment:
I think the idea is to pass meshes along with face data in DG communications and compute mortars on-the-fly in evolutions. However, for elliptic systems it would be nice to cache some things on mortars (Larry and I talked about this elsewhere). So I think it would be good to give every element the opportunity to update. Not sure how that works together with async AMR though.

Reply:
What about for h-refinement?  I guess we can do the mortar splitting or merging when we first receive messages, but that sounds more complicated than reorganizing as soon as we know what we need to do.  I don't think we need the mesh information, but I'm not entirely sure of that.

Reply:
Okay, I've changed this part of AdjustDomain to always pass the old mesh and element to the mutators, and removed the comment about how to do so in the future as it was confusing.   I chose to pass a `std::pair<Mesh,Element>` so that all three apply calls for the mutators take one additional argument to the db::mutate_apply call as this could allow a mutator to template the apply function on the additional argument in order to reduce code duplication.